File: constDemo/constDemo/ViewController.m
//
//  ViewController.m
//  constDemo
//
//  Created by NowOrNever on 26/06/2017.
//  Copyright © 2017 Focus. All rights reserved.
//

#import "ViewController.h"

@interface ViewController ()

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
//    [self setupUI];
    NSLog(@"%d",[self testWithNumber1:1 andNumber2:2]);
}

- (int)testWithNumber1:(int)number1 andNumber2:(int)number2{
    return number1 + number2;
}

- (void)setupUI{
    const int a = 6;
//    NSLog(@"%zd",a);
//    a = 7;    //  错误，常量
//    NSLog(@"%zd",a);
    
    int const b = 6;
//    b = 9; // error,常量
    
    int valueA = 6;
    int valueB = 8;
    const int *pointerA = &valueA;//    int const *pointerA = &valueA; 相等
    pointerA = &valueB;
//    *pointerA = 8; //error,pointerA指向一个数，该数不可变
    //指针可变，指针内容不可变
    
    int * const pointerC = &valueA;
    *pointerC = 9;
//    pointerC = &valueB; //error, * 在const左边，表明该指针不可变
    
//    int * const pointerC = &valueA;
//    修饰p则p的值不可改,p为指针，指针内容为地址，不可改
    
//    const 修饰 *p, 则*p内容不可改，指向的对象不可改


}


- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


@end

File: CAShapeLayerTest/CAShapeLayerTest/ViewController.swift
//
//  ViewController.swift
//  CAShapeLayerTest
//
//  Created by NowOrNever on 29/06/2017.
//  Copyright © 2017 Focus. All rights reserved.
//

import UIKit
import SnapKit

class ViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
//        setupUI()
        testDiy()
    }
    
    //抽成类
    fileprivate func testDiy(){
        let diyView = DiyView()
//        let diyView = DiyImageView.init(image: UIImage.init(named: "TifaLockhart"))
//        diyView.backgroundColor = UIColor.randomColor()
        view.addSubview(diyView)
        diyView.snp.makeConstraints { (make) in
            make.center.equalTo(view)
            make.height.equalTo(225)
            make.width.equalTo(300)
        }
    }
    
    
    //直接写，不抽成类
    fileprivate func setupUI(){
        let imageView = UIImageView.init(image: UIImage.init(named: "TifaLockhart"))
        view.addSubview(imageView)
        
        imageView.snp.makeConstraints { (make) in
            make.center.equalTo(view)
            make.height.equalTo(225)
            make.width.equalTo(300)
        }
        view.layoutIfNeeded()
        
        
        let path = UIBezierPath.init()
        path.move(to: CGPoint.init(x: 0, y: 0))
        path.addLine(to: CGPoint.init(x: 0, y: imageView.height))
        path.addLine(to: CGPoint.init(x: imageView.width, y: imageView.height))
        path.addLine(to: CGPoint.init(x: imageView.width, y: imageView.height * 0.5))
        path.close()
        
        let shapeLayer = CAShapeLayer.init()
        shapeLayer.path = path.cgPath
        
        imageView.layer.mask = shapeLayer
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }


}

